spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.java

```diff
@@ -732,6 +761,39 @@ protected void shutdownAndWaitOrCallback(@Nullable Runnable callback) {
 		}
 	}
 
+	/**
+	 * Wait for pending replies if a pending reply provider is configured.
+	 */
+	private void waitForPendingReplies() {
+		PendingReplyProvider provider = this.pendingReplyProvider;
+		if (provider != null && getShutdownTimeout() > 0) {
+			long deadline = System.currentTimeMillis() + getShutdownTimeout();
+			try {
+				while (isRunning() && System.currentTimeMillis() < deadline) {
+					int pendingCount = provider.getPendingReplyCount();
+					if (pendingCount <= 0) {
+						if (logger.isInfoEnabled()) {
+							logger.info("No pending replies detected, proceeding with shutdown.");
+						}
+						return;
+					}
+					if (logger.isInfoEnabled()) {
+						logger.info("Waiting for " + pendingCount + " pending replies before final shutdown...");
+					}
+					Thread.sleep(this.pendingReplyCheckInterval);
+				}
+				int remaining = provider.getPendingReplyCount();
+				if (remaining > 0) {
+					logger.warn("Shutdown timeout expired, but " + remaining + " pending replies still remain.");
+				}
+			}
+			catch (InterruptedException e) {
+				Thread.currentThread().interrupt();
+				logger.warn("Interrupted while waiting for pending replies.");
```

Comment:
I don't think we need so much logic here.
Currently, we utilize an `ActiveObjectCounter` for the consumers to be cancelled in this listener container.
I believe we can use it in the `RabbitTemplate` as well for those pending replies to be fulfilled.
Then in that `awaitShutdown`  hook we could take this `ActiveObjectCounter` from the `ListenerContainerAware` and wait on it as we do for consumers.